### PR TITLE
Support MNIST in ONNX

### PR DESCRIPTION
**Pull request by houseroad**
1) Fix bugs in MaxPool2d/MaxPool3d symbolic
2) Add symbolic for logsoftmax
3) Add symbolic for featuredropout

Now we can claim MNIST is supported.

### CODE REVIEW

File: torch/nn/_functions/dropout.py [hunk 1]
```diff
@@ -58,8 +58,9 @@ def backward(ctx, grad_output):
 class FeatureDropout(Dropout):
 
     @staticmethod
-    def symbolic(input, p=0.5, train=False, inplace=False):
-        return None
+    def symbolic(g, input, p=0.5, train=False, inplace=False):
+        # NB: No FeatureDropout in ONNX, use Dropout instead.
+        return Dropout.symbolic(g, input, p, train, inplace)
```
Comment (apaszke) [inline torch/nn/_functions/dropout.py]: Dropout and FeatureDropout are different. You’re silently exporting a wrong function

File: torch/nn/_functions/thnn/auto_symbolic.py [hunk 2]
```diff
@@ -26,6 +26,10 @@ def softmax_symbolic(g, input):
     return g.op('Softmax', input)
 
 
+def logsoftmax_symbolic(g, input):
+    return g.op("Log", g.op('Softmax', input).typeAs(input))
```
Comment (apaszke) [inline torch/nn/_functions/thnn/auto_symbolic.py]: Doesn’t ONNX have log softmax? That’s surprising. Softmax + Log is very unstable in backward

File: torch/nn/_functions/dropout.py (hunk 1 above)
Comment (houseroad) [reply torch/nn/_functions/dropout.py]: https://github.com/onnx/onnx/issues/113 We are discussing whether we should add FeatureDropout to ONNX. We should either add FeatureDropout or output some warning information and export FeatureDropout as a non-op.

File: torch/nn/_functions/thnn/auto_symbolic.py (hunk 2 above)
Comment (houseroad) [reply torch/nn/_functions/thnn/auto_symbolic.py]: Unfortunately, no logsoftmax in ONNX. For now, I think it's OK, because the exported model only runs inference. If we decide to enable backward, this may cause problem.

Comment (dzhulgakov) [reply torch/nn/_functions/thnn/auto_symbolic.py]: We should add LogSoftMax (sometimes called LogSumExp) to ONNX I think. It's pretty important for numerical stability, even in forward and most libraries implement it. Create an issue or PR?

**Review by dzhulgakov — COMMENTED** [review]
Btw, shall we create numerical tests for all ops we add or at least MNist model test?

File: torch/nn/_functions/dropout.py (hunk 1 above)
Comment (apaszke) [reply torch/nn/_functions/dropout.py]: If there’s a Bernoulli sampling op then you can implement feature dropout with sample + broadcast

File: torch/nn/_functions/thnn/auto_symbolic.py (hunk 2 above)
Comment (ezyang) [reply torch/nn/_functions/thnn/auto_symbolic.py]: Done in https://github.com/onnx/onnx/issues/115

File: torch/nn/_functions/dropout.py (hunk 1 above)
Comment (houseroad) [reply torch/nn/_functions/dropout.py]: No sampling op... We will see the results of the discussion.

Comment (apaszke) [reply torch/nn/_functions/dropout.py]: I think it's ok to merge it if it's used in Eval mode, but it should probably be an error or at least a warning in train mode.

Comment (houseroad) [reply torch/nn/_functions/dropout.py]: It's a good idea to raise an error if train=True.

**Comment by ezyang** [comment]
@dzhulgakov Yes, but you won't see it in this PR (they have to go in onnx-pytorch repo)

**Comment by ezyang** [comment]
@houseroad Don't forget to fix the FeatureDropout symbolic, so we can merge this

File: torch/nn/_functions/dropout.py [hunk 3]
```diff
@@ -58,8 +58,11 @@ def backward(ctx, grad_output):
 class FeatureDropout(Dropout):
 
     @staticmethod
-    def symbolic(input, p=0.5, train=False, inplace=False):
-        return None
+    def symbolic(g, input, p=0.5, train=False, inplace=False):
+        # NB: No FeatureDropout in ONNX, use Dropout instead.
+        if train:
+            raise ValueError("In train mode, FeatureDropout is different from Dropout.")
```
Comment (apaszke) [inline torch/nn/_functions/dropout.py]: Imagine you get this error when you export a model. You know it’s different, that’s why you used it instead. This error doesn’t help the user figure out what went wrong at all

Comment (houseroad) [reply torch/nn/_functions/dropout.py]: Yes, people may export the model in train mode and use it in inference mode, and vise versa. The message can still warn users when training flag is set true during exporting.

And this is just a temporary solution.
@ezyang has requested the FeatureDropout in https://github.com/onnx/onnx/issues/113.
We will update the exporter as long as it is introduced into ONNX.

Comment (fmassa) [reply torch/nn/_functions/dropout.py]: I think @apaszke meant that it would be better to change the error message to something like
```
onnx export for FeatureDropout not supported in train mode.
```

Comment (apaszke) [reply torch/nn/_functions/dropout.py]: Yes, that's exactly what I meant. People never use FeatureDropout directly, and telling that export failed because "FeatureDropout != Dropout" doesn't really carry any usable information for them

Comment (houseroad) [reply torch/nn/_functions/dropout.py]: @fmassa Thanks for the explanation. @apaszke Make sense. I will update the error message.
